category/bifunctor.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2018 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+
+Functors with two arguments
+-/
+
+import data.sum
+       category.basic category.functor
+       tactic.basic
+
+universes u₀ u₁ u₂ v₀ v₁ v₂
+
+class bifunctor (F : Type u₀ → Type u₁ → Type u₂) :=
+(bimap : Π {α α' β β'}, (α → α') → (β → β') → F α β → F α' β')
+export bifunctor ( bimap )
+
+class is_lawful_bifunctor (F : Type u₀ → Type u₁ → Type u₂) [bifunctor F] :=
+(id_bimap : Π {α β} (x : F α β), bimap id id x = x)
+(comp_bimap : Π {α₀ α₁ α₂ β₀ β₁ β₂} (f : α₀ → α₁) (f' : α₁ → α₂)
+  (g : β₀ → β₁) (g' : β₁ → β₂) (x : F α₀ β₀),
+  bimap f' g' (bimap f g x) = bimap (f' ∘ f) (g' ∘ g) x)
+  -- `comp_bimap` constrasts with `comp_map` from lawful functors but
+  -- this form is more easily usable in a set of rewrite rules.
+export is_lawful_bifunctor (id_bimap comp_bimap)
+
+attribute [higher_order bimap_id_id] id_bimap
+attribute [higher_order bimap_comp_bimap] comp_bimap
+
+export is_lawful_bifunctor (bimap_id_id bimap_comp_bimap)
+variables {F : Type u₀ → Type u₁ → Type u₂} [bifunctor F]
+
+namespace bifunctor
+
+@[reducible]
+def first {α α' β} (f : α → α') : F α β → F α' β :=
+bimap f id
+
+@[reducible]
+def second {α β β'} (f : β → β') : F α β → F α β' :=
+bimap id f
```

Comment:
Are you calling these `first` and `second` instead of `fst` and `snd` due to a name conflict? If not I would stick to `fst` and `snd` for consistency.

Reply:
I'm mimicking Haskell's naming scheme

Reply:
I'll defer to your judgement then.